### PR TITLE
Read delegate reward ratio from IoTeX chain 246

### PR DESCRIPTION
**Pull request by lzxm160**

### CODE REVIEW

File: indexprotocol/votings/stakingprotocol.go [hunk 1]
```diff
@@ -275,6 +292,64 @@ func (p *Protocol) getStakingBucketInfoByEpoch(height, epochNum uint64, delegate
...
 }
 
+func (p *Protocol) getAllStakingDelegateRewardPortions(epochStartHeight, epochNumber uint64, chainClient iotexapi.APIServiceClient) (blockRewardPercentage, epochRewardPercentage, foundationBonusPercentage map[string]float64, err error) {
+	blockRewardPercentage = make(map[string]float64)
+	epochRewardPercentage = make(map[string]float64)
+	foundationBonusPercentage = make(map[string]float64)
+	if epochStartHeight == p.epochCtx.FairbankHeight() {
+		// init from contract,from contract deployed height to epochStartheight-1,get latest portion
+		if p.rewardPortionContract == "" {
+			// todo make sure if ignore this error
```
Comment (lzxm160) [inline indexprotocol/votings/stakingprotocol.go:302]: need to decide if return error or ignore it

File: indexprotocol/votings/stakingprotocol.go [hunk 2]
```diff
@@ -275,6 +292,64 @@ func (p *Protocol) getStakingBucketInfoByEpoch(height, epochNum uint64, delegate
...
+			//err = errors.New("portion contract address is empty")
+			return
+		}
+		if p.rewardPortionContractDeployHeight > epochStartHeight {
+			// todo make sure if ignore this error
```
Comment (lzxm160) [inline indexprotocol/votings/stakingprotocol.go:307]: ditto

File: indexprotocol/votings/stakingprotocol.go [hunk 3]
```diff
@@ -275,6 +292,64 @@ func (p *Protocol) getStakingBucketInfoByEpoch(height, epochNum uint64, delegate
...
+		blockRewardPercentage, epochRewardPercentage, foundationBonusPercentage, err = getlog(p.rewardPortionContract, p.rewardPortionContractDeployHeight, count, chainClient, p.abi)
+		if err != nil {
+			err = errors.Wrap(err, "get log from chain error")
+			return
+		}
+	} else {
+		// get from mysql first
+		blockRewardPercentage, epochRewardPercentage, foundationBonusPercentage, err = getLastEpochPortion(p.Store.GetDB(), epochNumber-1)
+		if err != nil && errors.Cause(err) != indexprotocol.ErrNotExist {
+			// todo make sure if ignore this error
```
Comment (lzxm160) [inline indexprotocol/votings/stakingprotocol.go:321]: ditto

Comment (lizhefeng) [reply indexprotocol/votings/stakingprotocol.go:321]: I don't think we need to check errors.Cause(err) != indexprotocol.ErrNotExist here. If errors.Cause(err) = indexprotocol.ErrNotExist, it is unexpected and we should throw the error, too.

File: config.yaml [hunk 4]
```diff
@@ -7,6 +7,8 @@ dardanellesOn: true
 fairbankHeight: 5165641
 consensusScheme: "ROLLDPOS"
 rangeQueryLimit: 100
+rewardPortionContract: "io16dxewjaec7ddxuk8n6g2dpezthzjlfuqu4w9df"
```
Comment (lizhefeng) [inline config.yaml:10]: Use mainnet address: io1lfl4ppn2c3wcft04f0rk0jy9lyn4pcjcm7638u

File: config.yaml [hunk 5]
```diff
@@ -7,6 +7,8 @@ dardanellesOn: true
...
+rewardportionContractDeployHeight: 3615690
```
Comment (lizhefeng) [inline config.yaml:11]: Use mainnet deploy height: 5095225

File: indexprotocol/votings/protocol.go [hunk 6]
```diff
@@ -76,13 +77,13 @@ const (
 	createVotingMetaTable = "CREATE TABLE IF NOT EXISTS %s (epoch_number DECIMAL(65, 0) NOT NULL, " +
 		"voted_token DECIMAL(65,0) NOT NULL, delegate_count DECIMAL(65,0) NOT NULL, total_weighted DECIMAL(65, 0) NOT NULL, " +
 		"UNIQUE KEY %s (epoch_number))"
-	selectVotingResult = "SELECT * FROM %s WHERE epoch_number=? AND delegate_name=?"
-	insertVotingResult = "INSERT INTO %s (epoch_number, delegate_name, operator_address, reward_address, " +
+	selectVotingResult                   = "SELECT * FROM %s WHERE epoch_number=? AND delegate_name=?"
+	selectVotingResultFromStakingAddress = "SELECT * FROM %s WHERE epoch_number=?"
```
Comment (lizhefeng) [inline indexprotocol/votings/protocol.go:81]: selectVotingResultForAllDelegates

File: indexprotocol/votings/stakingprotocol.go [hunk 7]
```diff
@@ -73,15 +85,20 @@ func (p *Protocol) updateStakingResult(tx *sql.Tx, candidates *iotextypes.Candid
...
 		if err != nil {
 			return errors.Wrap(err, "failed to convert IoTeX address to ETH address")
 		}
-		blockRewardPortion, epochRewardPortion, foundationBonusPortion, err := p.getDelegateRewardPortions(stakingAddress, gravityHeight)
-		if err != nil {
-			return errors.Errorf("get delegate reward portions:%s,%d,%s", stakingAddress.String(), gravityHeight, err.Error())
-		}
+		blockRewardPortion = blockRewardPortionMap[strings.ToLower(stakingAddress.String()[2:])]
+		epochRewardPortion = epochRewardPortionMap[strings.ToLower(stakingAddress.String()[2:])]
+		foundationBonusPortion = foundationBonusPortionMap[strings.ToLower(stakingAddress.String()[2:])]
```
[the remark this answers is not in the document]
Comment (lzxm160) [reply indexprotocol/votings/stakingprotocol.go:101]: sure

File: indexprotocol/votings/stakingprotocol.go [hunk 8]
```diff
@@ -412,16 +407,15 @@ func ownerAddressToNameMap(candidates *iotextypes.CandidateListV2) (ret map[stri
 	return
 }
 
-func getlog(contractAddress string, from, count uint64, chainClient iotexapi.APIServiceClient, delegateABI abi.ABI) (blockReward, epochReward, foundationReward map[string]float64, err error) {
+func getlog(contractAddress string, from, count uint64, chainClient iotexapi.APIServiceClient, delegateProfileABI abi.ABI) (blockReward, epochReward, foundationReward map[string]float64, err error) {
```
Comment (lizhefeng) [inline indexprotocol/votings/stakingprotocol.go:410]: getLog

File: indexprotocol/votings/protocol.go [hunk 9]
```diff
@@ -137,27 +137,26 @@ type (
...
+	timeTableOperator             *committee.TimeTableOperator
+	epochCtx                      *epochctx.EpochCtx
+	GravityChainCfg               indexprotocol.GravityChain
+	voteCfg                       indexprotocol.VoteWeightCalConsts
+	SkipManifiedCandidate         bool
+	VoteThreshold                 *big.Int
+	ScoreThreshold                *big.Int
+	SelfStakingThreshold          *big.Int
+	rewardPortionCfg              indexprotocol.RewardPortionCfg
+	abi                           abi.ABI
```
Comment (lizhefeng) [inline indexprotocol/votings/protocol.go:155]: we don't need the abi as a field in protocol. We can construct it in getLog() instead.

Comment (lzxm160) [reply indexprotocol/votings/protocol.go:155]: sure

**Pull request merged**